tests/lib/tools/tests.session

```diff
@@ -231,12 +231,64 @@ main() {
 	unit_name="tests.session-$uuid.service"
 	tmp_dir="$TESTSTMP/tests.session-$uuid"
 	mkdir "$tmp_dir"
-	mkfifo -m 0666 "$tmp_dir/dbus-monitor.pipe" "$tmp_dir/ready.pipe" "$tmp_dir/result.pipe" "$tmp_dir/stdin.pipe" "$tmp_dir/stdout.pipe" "$tmp_dir/stderr.pipe"
+	mkfifo -m 0666 "$tmp_dir/result.pipe" "$tmp_dir/stdin.pipe" "$tmp_dir/stdout.pipe" "$tmp_dir/stderr.pipe"
 	trap 'rm -rf $tmp_dir' EXIT
 
-	# Run dbus-monitor waiting for systemd JobRemoved signal. Process the output
-	# with awk, forwarding the result of the service to a fifo.
-	monitor_expr="type='signal', sender='org.freedesktop.systemd1', interface='org.freedesktop.systemd1.Manager', path='/org/freedesktop/systemd1', member='JobRemoved'"
+	# Use busctl to spawn a command. The command is wrapped in shell, runuser -l
+	# and redirects to capture output. Sadly busctl doesn't support passing file
+	# descriptors https://github.com/systemd/systemd/issues/14954 As a workaround
+	# we pass a set of pipes. This is good for non-interactive work.
+	# NOTE: /dev/stdin is provided explicitly to trigger special behavior in bash,
+	# as otherwise background tasks are started with /dev/null for stdin.
+	cat </dev/stdin >"$tmp_dir/stdin.pipe" &
+	cat_stdin_pid=$!
+	cat <"$tmp_dir/stdout.pipe" >&1 &
+	cat_stdout_pid=$!
+	cat <"$tmp_dir/stderr.pipe" >&2 &
+	cat_stderr_pid=$!
+
+	(
+		echo "#!/bin/sh"
+		test -n "$pid_file" && echo "echo \$$ >\"$pid_file\""
+		printf "exec "
+		for arg in "$@"; do
+			printf '%q ' "$arg"
+		done
+		echo
+	)>"$tmp_dir/exec"
+	chmod +x "$tmp_dir/exec"
+
+	# When systemd version is higher than 252 we use systemd-run to execute
+	# In other scenarios we use the older approach: bustctl with the dbus monitor
+	if [ "$(systemctl --version | awk '/systemd [0-9]+/ { print $2 }')" -gt 252 ]; then
+  		systemd-run \
+  		--unit="$unit_name" \
+		--description "tests.session running $* as $user" \
+		--quiet \
+		--wait \
```

Comment:
With `--wait`, if the service fails, the call to systemd-run fails. If the service fails, because the shebang is `#!/bin/bash -e`, the script will fail. So the lines after will not happen and the failed service will not be cleaned (`systemctl reset-failed`).

So we should either have:
 * call `systemd-run` with `--wait --collect` and without `--unit="${unit_name}"`. And we remove the calls to `systemctl cat` and `systemctl reset-failed`after.
 * call `systemd-run` with `--wait --unit="${unit_name}"` and without `--collect`. And we keep the calls to `systemctl cat` and `systemctl reset-failed`after.

Reply:
We are missing `|| true`.